### PR TITLE
[string] Avoid dyncasts through direct metatype comparison.

### PR DESCRIPTION
**Pull request by milseman**
Compare directly against a handful of final classes, for modest wins.

A small step for https://bugs.swift.org/browse/SR-6823

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/StringGuts.swift [hunk 1]
```diff
@@ -386,13 +386,16 @@ extension _StringGuts {
   }
 
   @_versioned
-  @inline(never) // Hide CF dependency
+  @_inlineable
+  //@inline(never) // Hide CF dependency
```
Comment (jrose-apple) [inline stdlib/public/core/StringGuts.swift]: Why is this safe to comment out?

Comment (milseman) [reply stdlib/public/core/StringGuts.swift]: Ah, this was part of a different optimization attempt I was making. It's safe to comment out as the reference to CFStringGetLength was changed to a different inline(never) call. I should just drop this from this PR though. It's more so an attempt to expose the triviality of ARC.

Comment (milseman) [reply stdlib/public/core/StringGuts.swift]: Ah, this was part of a different optimization attempt I was making. It's safe to comment out as the reference to CFStringGetLength was changed to a different inline(never) call. I should just drop this from this PR though. It's more so an attempt to expose the triviality of ARC.

**Comment by milseman** [comment]
@swift-ci please smoke test

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
(Expecting a modest few % gains here on bridging, nothing monumental)

**Comment by milseman** [comment]
@shahmishal @jrose-apple this has been running for 6.5 hours, is that expected?

https://ci.swift.org/job/swift-PR-osx-smoke-perf/180/

**Comment by jrose-apple** [comment]
Some of the other benchmark bots have been timing out. I wonder if we broke them some time this week.

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Pull request closed without merging**